synapse/http/site.py

```diff
@@ -334,6 +356,9 @@ def _finished_processing(self):
         except Exception as e:
             logger.warning("Failed to stop metrics: %r", e)
 
+        # Break the cycle between SynapseRequest and LoggingContext.
+        self.logcontext = None
```

Comment:
I don't remember why I felt it was necessary to manually break this, but it seems plausible.

Reply:
Hmm, I'm a bit concerned that we do `assert self.logcontext` in various places and its a bit non-obvious to me that they will never get called after this point.

Reply:
This PR only adds one of these, I think, to `finish`, which immediately calls `_finished_processing` where the other one is.

I believe that `_finished_processing` can be called once (and only once), so this should be safe.

Reply:
Alternately -- I suppose I could just run in the sentinel context then or something? Or do you have suggestions of what to do?

Reply:
I'd rather it exploded than accidentally record stuff against the sentinel context. Ideally our log contexts would be robust enough where we could trust tat we could just do `current_context()`. Maybe instead of having a pointer to the request in the log context we should store the actual values we care about? It's more space but a) no circular dependency and b) we don't recalculate the values every time we log something?

Reply:
> Maybe instead of having a pointer to the request in the log context we should store the actual values we care about? It's more space but a) no circular dependency and b) we don't recalculate the values every time we log something?

I think that would work, the annoying bit becomes that there will be a couple of places that those need to be listed (since we have to handle the child-parent context relationship). I'll think a bit more about it!

Reply:
We could have a `RequestMetadata` struct that we point to from various places? I don't know. I'm not a massive fan of the current state of affairs but I don't think we should block this PR on it too much

Reply:
I had the same though last night. 👍

Reply:
I made this change, but it got a little funky since `requester` gets set later, so I needed to make it a `setter`...

Reply:
Thanks!